cmd/root.go

```diff
@@ -86,7 +96,21 @@ func (c *rootCommand) execute() {
 	exitCode := -1
 	defer func() {
 		cancel()
-		c.stopLoggers()
+
+		// Wait for loggers and tracer provider to stop.
+		// As both methods already use a timeout, just wait.
+		var wg sync.WaitGroup
+		ff := [...]func(){c.stopLoggers, c.stopTracerProvider}
+		wg.Add(len(ff))
+		for _, f := range ff {
+			f := f
+			go func() {
+				f()
```

Comment:
nit: I would make `stopLoggers(...)` and `stopTracerProvider(...)` methods accept a `context.Context` parameter, as you seem to have a more "global context" available. If the global context gets canceled, it won't impact the other goroutines.

Reply:
If I'm understanding you correctly, the idea of making both `stopLoggers()` and `stopTracerProvider()` accept a `context` LGTM. So it would be similar to:
<details>
<summary>code</summary>

```go
func (c *rootCommand) stopLoggers(ctx context.Context) {
	done := make(chan struct{})
	go func() {
		c.loggersWg.Wait()
		close(done)
	}()
	close(c.stopLoggersCh)
	select {
	case <-done:
	case <-ctx.Done():
		c.globalState.FallbackLogger.Error("The logger didn't stop in time")
	}
}

func (c *rootCommand) stopTracerProvider(ctx context.Context) {
	if err := c.globalState.TracerProvider.Shutdown(ctx); err != nil {
		c.globalState.FallbackLogger.Errorf(
			"The tracer provider didn't stop gracefully: %v", err,
		)
	}
}
```
</details>

And then from `execute()`'s defer we could use a single context with timeout such as:
<details>
<summary>code</summary>

```go
var wg sync.WaitGroup
ctx, cancel := context.WithTimeout(context.Background(), waitLoggerTracerProviderStopTimeout)
defer cancel()
ff := [...]func(context.Context){c.stopLoggers, c.stopTracerProvider}
wg.Add(len(ff))
for _, f := range ff {
	f := f
	go func() {
		f(ctx)
		wg.Done()
	}()
}
wg.Wait()
```
</details>

Can you clarify what do you  mean with _"you seem to have a more 'global context' available"_?
Are you referring to the `c.globalState.Ctx`? For what I understand we are in control of that context from `execute()` and [canceling](https://github.com/grafana/k6/blob/bbc461bb4d598493d7df16c01c95f965ad2daed0/cmd/root.go#L98) it just before stopping loggers and tracer provider, so I think it should be OK to build the context with timeout off of `Background()`?

Reply:
Regarding code snippets - yeah, this is what I had in mind. In general, it's a good practise to accept `context.Context` in shutdown functions that can be blocked. This pattern is widely used throughout stdlib ([example](https://cs.opensource.google/go/go/+/master:src/net/http/server.go;l=2979))

> Can you clarify what do you mean with "you seem to have a more 'global context' available"?
Are you referring to the c.globalState.Ctx? For what I understand we are in control of that context from execute() and [canceling](https://github.com/grafana/k6/blob/bbc461bb4d598493d7df16c01c95f965ad2daed0/cmd/root.go#L98) it just before stopping loggers and tracer provider, so I think it should be OK to build the context with timeout off of Background()?

Yes, I was referring to that context. The question is - Is there a use case where that global context gets canceled, but you don't want to shut down loggers and tracers? If not, I would probably derive the context you inject to shutdown functions from that global context and ensure that the "green path" executes your shutdown logic before the global context gets canceled. This would ensure that no matter what goes wrong, you can always shutdown these goroutines by canceling the global context.

Reply:
This discussion is superseded by the discussion held in https://github.com/grafana/k6/pull/3445#discussion_r1389525664, and particularly by commit [3fbc0b8](https://github.com/grafana/k6/pull/3445/commits/3fbc0b89391478708a53973701f64cfec7e509de).

Reply:
I will resolve this at this point